src/stm32f411/dma.cpp

```diff
@@ -0,0 +1,250 @@
+  hal::cortex_m::enable_interrupt(irq::dma2_channel4, handle_dma_interrupt<12>);
+  hal::cortex_m::enable_interrupt(irq::dma2_channel5, handle_dma_interrupt<13>);
+  hal::cortex_m::enable_interrupt(irq::dma2_channel6, handle_dma_interrupt<14>);
+  hal::cortex_m::enable_interrupt(irq::dma2_channel7, handle_dma_interrupt<15>);
+}
+
+hal::atomic_spin_lock dma_spin_lock;
+hal::basic_lock* dma_lock = &dma_spin_lock;
+
+// This is here for cost sizing
+[[maybe_unused]] constexpr auto dma_memory_usage_size =
+  sizeof(dma_callbacks) + sizeof(dma_spin_lock) + sizeof(dma_lock);  // NOLINT
+}  // namespace
+
+void set_dma_lock(hal::basic_lock& p_lock)
+{
+  dma_lock = &p_lock;
+}
+
+dma_channel_stream_t set_available_stream(
+  peripheral p_dma,
+  std::span<dma_channel_stream_t> const& p_possible_streams)
+{
+  auto dma_addr = get_dma_reg(p_dma);
+  while (true) {
+    for (auto stream_candidate : p_possible_streams) {
+      if (bit_extract<dma_stream_config::channel_select>(
+            dma_addr->stream[stream_candidate.stream].configure) ==
+          stream_candidate.channel) {
+        return stream_candidate;
+      } else if (bit_extract<dma_stream_config::stream_enable>(
+                   dma_addr->stream[stream_candidate.stream].configure) == 0) {
+        bit_modify(dma_addr->stream[stream_candidate.stream].configure)
+          .insert(dma_stream_config::channel_select, stream_candidate.channel);
+        return stream_candidate;
+      }
+    }
+  }
+}
+
+dma_channel_stream_t setup_dma_transfer(
+  peripheral p_dma,
+  std::span<dma_channel_stream_t> const& p_possible_streams,
+  dma_settings_t const& p_configuration,
+  hal::callback<void()> p_interrupt_callback)
+{
+  auto dma_addr = get_dma_reg(p_dma);
+
+  std::lock_guard take_dma_lock(*dma_lock);
+  initialize_dma(p_dma);
+
+  auto selected_config = set_available_stream(p_dma, p_possible_streams);
+
+  switch (p_configuration.transfer_type) {
+    case dma_transfer_type::peripheral_to_memory:
+      dma_addr->stream[selected_config.stream].peripheral_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.source);
+      dma_addr->stream[selected_config.stream].memory_0_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.destination);
+      break;
+    case dma_transfer_type::memory_to_peripheral:
+      dma_addr->stream[selected_config.stream].peripheral_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.destination);
+      dma_addr->stream[selected_config.stream].memory_0_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.source);
+      break;
+    default:
+      dma_addr->stream[selected_config.stream].peripheral_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.source);
+      dma_addr->stream[selected_config.stream].memory_0_address =
+        std::bit_cast<std::uintptr_t>(p_configuration.destination);
+  }
+
+  dma_addr->stream[selected_config.stream].transfer_count =
+    p_configuration.transfer_length;
```

Comment:
```suggestion

  auto& stream = dma_addr->stream[selected_config.stream];
  switch (p_configuration.transfer_type) {
    case dma_transfer_type::peripheral_to_memory:
      stream.peripheral_address = std::bit_cast<std::uintptr_t>(p_configuration.source);
      stream.memory_0_address = std::bit_cast<std::uintptr_t>(p_configuration.destination);
      break;
    case dma_transfer_type::memory_to_peripheral:
      stream.peripheral_address = std::bit_cast<std::uintptr_t>(p_configuration.destination);
      stream.memory_0_address = std::bit_cast<std::uintptr_t>(p_configuration.source);
      break;
    default:
      stream.peripheral_address = std::bit_cast<std::uintptr_t>(p_configuration.source);
      stream.memory_0_address = std::bit_cast<std::uintptr_t>(p_configuration.destination);
  }

  stream.transfer_count = p_configuration.transfer_length;
```

The compiler is smart enough to optimize away the use of `auto` references, so you can use them for naming shorter versions of something.